File: Algorithms/Sorting/ShellSort/ShellSort.c
#include <stdio.h>
#include <stdlib.h>

/* Prototypes */
void shell_sort(int *arr, int N);
void swap(int *arr, int i, int j);
void show_items(int *arr, int N);

int main(void)
{
    int N = 20;
    int *test1 = malloc(sizeof(int) * N);
    int *test2 = malloc(sizeof(int) * N);
    for (int i = 0; i < N; i++)
    {
        int rand1 = rand() * .000001;
        int rand2 = rand() * .000001;
        test1[i] = rand1;
        test2[i] = rand2;
    }
    printf("Arrays before sort:\n");
    show_items(test1, N);
    show_items(test2, N);
    shell_sort(test1, N);
    shell_sort(test2, N);
    printf("Arrays after sort:\n");
    show_items(test1, N);
    show_items(test2, N);
    free(test1);
    free(test2);
}

void shell_sort(int *arr, int N)
{
    // First find the start of the increment sequence 3x + 1
    int H = 1;
    while (H < N / 3)
    {
        H = 3 * H + 1;
    }
    while (H >= 1)
    {
        // Now we do insertion sort based on the decrement sequence of j -= H
        for (int i = H; i < N; i++)
        {
            for (int j = i; j >= H; j -= H)
            {
                if (arr[j] < arr[j - H])
                {
                    swap(arr, j, j - H);
                }
                else { break; }
            }
        }
        // decrement H
        H = H / 3;
    }
}

void swap(int *arr, int i, int j)
{
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

void show_items(int *arr, int N)
{
    printf("[");
    for (int i = 0; i < N; i++)
    {
        if (i == N - 1)
        {
            printf("%i]\n", arr[i]);
            break;
        }
        printf("%i, ", arr[i]);
    }
}

File: Algorithms/Sorting/ShellSort/ShellSort.js
class ShellSort {

    static sort(arr) {
        const N = arr.length
        let H = 1
        while (H < Math.floor(N / 3)) {
            H = 3 * H + 1
        }
        while (H >= 1) {
            for (let i = H; i < N; i++) {
                for (let j = i; j >= H; j -= H) {
                    if (arr[j] < arr[j - H]) {
                        ShellSort.swap(arr, j, j - H)
                    } else {
                        break;
                    }
                }
            }
            H = Math.floor(H / 3)
        }
    }

    static swap(arr, x, y) {
        const temp = arr[x]
        arr[x] = arr[y]
        arr[y] = temp
    }

    static main() {
        const test1 = Array(20)
        const test2 = Array(20)
        for (let i = 0; i < 20; i++) {
            test1[i] = Math.floor(Math.random() * 100)
            test2[i] = Math.floor(Math.random() * 100)
        }
        console.log("Arrays before the sort:")
        console.log(test1)
        console.log(test2)
        ShellSort.sort(test1)
        ShellSort.sort(test2)
        console.log("Arrays after the sort:")
        console.log(test1)
        console.log(test2)
    }
}

ShellSort.main()